packages/analytics/index.test.ts

```diff
@@ -34,51 +34,144 @@ import { GtagCommand, EventName } from './src/constants';
 import { findGtagScriptOnPage } from './src/helpers';
 import { removeGtagScript } from './testing/gtag-script-util';
 import { Deferred } from '@firebase/util';
+import { AnalyticsError } from './src/errors';
 
 let analyticsInstance: FirebaseAnalytics = {} as FirebaseAnalytics;
-const analyticsId = 'abcd-efgh';
-const gtagStub: SinonStub = stub();
+const fakeMeasurementId = 'abcd-efgh';
+const fakeAppParams = { appId: 'abcdefgh12345:23405', apiKey: 'AAbbCCdd12345' };
+let fetchStub: SinonStub = stub();
 const customGtagName = 'customGtag';
 const customDataLayerName = 'customDataLayer';
 
-describe('FirebaseAnalytics instance tests', () => {
-  it('Throws if no analyticsId in config', () => {
-    const app = getFakeApp();
-    const installations = getFakeInstallations();
-    expect(() => analyticsFactory(app, installations)).to.throw(
-      'field is empty'
-    );
+function stubFetch(status: number, body: object): void {
+  fetchStub = stub(window, 'fetch');
+  const mockResponse = new Response(JSON.stringify(body), {
+    status
   });
-  it('Throws if creating an instance with already-used analytics ID', () => {
-    const app = getFakeApp(analyticsId);
-    const installations = getFakeInstallations();
-    resetGlobalVars(false, { [analyticsId]: Promise.resolve() });
-    expect(() => analyticsFactory(app, installations)).to.throw(
-      'already exists'
-    );
+  fetchStub.returns(Promise.resolve(mockResponse));
+}
+
+describe('FirebaseAnalytics instance tests', () => {
+  describe('Initialization', () => {
+    beforeEach(() => resetGlobalVars());
+
+    it('Throws if no appId in config', () => {
+      const app = getFakeApp({ apiKey: fakeAppParams.apiKey });
+      const installations = getFakeInstallations();
+      expect(() => analyticsFactory(app, installations)).to.throw(
+        AnalyticsError.NO_APP_ID
+      );
+    });
+    it('Throws if no apiKey or measurementId in config', () => {
+      const app = getFakeApp({ appId: fakeAppParams.appId });
+      const installations = getFakeInstallations();
+      expect(() => analyticsFactory(app, installations)).to.throw(
+        AnalyticsError.NO_API_KEY
+      );
+    });
+    it('Warns if config has no apiKey but does have a measurementId', () => {
+      stubFetch(200, { measurementId: fakeMeasurementId });
```

Comment:
Do you need to stub fetch if there is no apiKey? 
In the expect statement later, can we match the full error string (or match a longer string) to make it easier to understand? e.g. `apiKey is not available, falling back to 'fakeMeasurementId'`

Reply:
Done.